### DISCUSSION TITLE
Exception in BaseEnhancement

### DISCUSSION THREAD

**Discussion by markus-nclose** [jertel/elastalert2#516] [Q&A] [answered]
Hi All.

We have a ```try``` inside the ```process``` function in our Enhancement. (see below example snippet). The ```raise DropMatchException()``` is not called inside the ```try```. Are we missing something obvious?

```
class MyEnhancement(BaseEnhancement):

    def process(self, match):
        try:
            raise DropMatchException()
        except Exception as e:
            print('bugger')
```

Thanks

**Comment by jertel** [comment, marked as answer]
Do you have any errors in your ElastAlert 2 log/output that might indicate a problem?

Does your `my_enhancements.py` file have the correct import statement at the top?
```
from elastalert.enhancements import BaseEnhancement
```

Does your rule have the required enhancement configuration?
```
match_enhancements:
- "mymodules.my_enhancements.MyEnhancement"
```

Try adding the following `print` to the MyEnhancement class, between the process() function signature and the `try:`:
```
    def process(self, match):
        print('******************************************************\n')
        try:
```
Then review your ElastAlert 2 console output (or Docker logs) and you should see those asterisks.

**Reply by markus-nclose** [reply]
We did quite extensive testing on this. 

We have the correct import - see comment below
We have the correct enhancement configuration - we know this because if you move the ```raise DropMatchException()``` above the ```try```, it drops the event.
We did try the asterisks - we use ```logger.info('*'*100)```. We can see the function execute, we also added a print right before the ```DropMatchException``` to verify it running. It simply just skips the DropMatchException.

**Reply by jertel** [reply]
I took your exact enhancement class, added the import to the top, added the println, configured it in my rule and I see this output:

```
2021-10-21 12:25:01,247    DEBUG        elasticsearch < {"_index":"elastalert_status_silence","_type":"_doc","_id":"zfTPonwBFE2TZycz0ZH5","_version":1,"result":"created","_shards":{"total":2,"successful":1,"failed":0},"_seq_no":0,"_primary_term":6}
******************************************************

bugger
2021-10-21 12:25:02,326     INFO           elastalert Sent email to ['test@test.invalid']
```

**Reply by jertel** [reply]
Maybe it's a misunderstanding of try/except -- Are you wanting the event to be dropped while inside the try? If so, that's not going to happen since the `except` clause is catching it and preventing it from bubbling up.

**Reply by VaderZA** [reply]
@jertel thanks for looking into this.
Our enhancement is contacting an external system which is why we using `try`
Our enhancement is working, we just came across this behaviour:

I am going to skip the basics so assume all imports and enhancement requirements are in place, here is a **simplified example**

## This works
```python
class MyEnhancement(BaseEnhancement):

    def process(self, match):
        if self.test_alert(match):
            raise DropMatchException()

    def test_alert(self, match):
        try:
            # this is where we contact and external system
            return True
        except Exception as e:
            print(f"An error occurred: {e}")
            return False
```
---
## This works
```python
class MyEnhancement(BaseEnhancement):
    
    def process(self, match):
        raise DropMatchException()
        try:
            print("A stupid example")
        except Exception as e:
            print(f"An error occured: {e}")
```
---

## This does not work
```python
class MyEnhancement(BaseEnhancement):
    
    def process(self, match):
        try:
            # this is where we contact and external system
            raise DropMatchException()
        except Exception as e:
            print(f"An error occurred: {e}")
```
If the `raise DropMatchException()` is in a try statement of the process function the alert would not be dropped.
There were no errors at all, it was as though the `raise DropMatchException()` was doing nothing.

**Comment by VaderZA** [comment]
@jertel 
I just re-read your last post and totally understand. Thanks again, please close.
